File: Project2/WorldSimulator/World.java
package WorldSimulator;

import java.util.*;

public abstract class World
{

    private final Vector2 size;
    private final Vector<Organism> organisms = new Vector<>();
    private int turnNumber = 0;
    private boolean playerAlive = true;
    private final ArrayList<String> logs = new ArrayList<>();

    public World(int width, int height)
    {
        size = new Vector2(width, height);
    }


    public void AddOrganism(Organism org)
    {
        org.setWorld(this);
        organisms.add(org);
    }

    public void RemoveOrganism(Organism org)
    {
        for (Organism organism : organisms)
        {
            if (organism == org)
            {
                organisms.remove(organism);
                break;
            }
        }
    }

    public Organism GetOrganism(Vector2 pos)
    {
        for (Organism organism : organisms)
        {
            if (pos.equals(organism.getPos()))
                return organism;
        }
        return null;
    }

    public Vector<Organism> GetOrganisms()
    {
        return organisms;
    }

    public Vector2 GetSize()
    {
        return size;
    }

    public int GetTurnNumber()
    {
        return turnNumber;
    }
    public void SetTurnNumber(int value)
    {
        turnNumber = value;
    }

    public void AddLog(String str)
    {
        logs.add(str);
    }

    public boolean IsPlayerAlive()
    {
        return playerAlive;
    }

    public void SetPlayerAlive(boolean value)
    {
        playerAlive = value;
    }

    public boolean ContainsPos(Vector2 pos)
    {
        return pos.x >= 0 && pos.x < size.x&& pos.y >= 0 && pos.y < size.y;
    }

    public Vector2 GetRandomEmptyTile()
    {
        Vector2 pos = new Vector2();
        Random rand = new Random();

        int safetyLimit = 1000;
        for (int i = 0; i < safetyLimit; i++)
        {
            pos.x = rand.nextInt(size.x);
            pos.y = rand.nextInt(size.y);
            // if there is no organism on the chosen tile, return the position
            if (GetOrganism(pos) == null)
                break;
        }

        return pos;
    }

    public abstract ArrayList<Vector2> GetNeighbors(Vector2 v);
    public abstract ArrayList<Vector2> GetNeighborsWithDiagonals(Vector2 v);
    public Vector2 GetRandomNeighbor(Vector2 pos)
    {
        Random rand = new Random();
        ArrayList<Vector2> neighbors = GetNeighbors(pos);
        return neighbors.get(rand.nextInt(neighbors.size()));
    }
    public boolean HasEmptyNeighbor(Vector2 pos)
    {
        ArrayList<Vector2> neighbors = GetNeighbors(pos);
        for (Vector2 n : neighbors)
        {
            if (ContainsPos(n) && GetOrganism(n) == null)
                return true;
        }
        return false;
    }
    public Vector2 GetRandomEmptyNeighbor(Vector2 pos)
    {
        Vector<Vector2> targets = new Vector<>();

        ArrayList<Vector2> neighbors = GetNeighbors(pos);
        for (Vector2 n : neighbors)
        {
            if (ContainsPos(n) && GetOrganism(n) == null)
                targets.add(n);
        }

        if (targets.size() > 0)
        {
            Random rand = new Random();
            return targets.get(rand.nextInt(targets.size()));
        }
        else
            return null;
    }

    public void ExecuteTurn()
    {
        // order organisms by their priority
        organisms.sort((Organism org1, Organism org2) -> org2.getPriority() - org1.getPriority());

        // execute actions
        int orgCount = organisms.size();
        for (int i = 0; i < orgCount; i++)
        {
            if (!organisms.get(i).isAlive())
                continue;

            organisms.get(i).Action();

            // handle collisions
            for (int j = 0; j < organisms.size(); j++)
            {
                if (i != j && organisms.get(j).isAlive() && organisms.get(i).getPos().equals(organisms.get(j).getPos()))
                {
                    organisms.get(i).Collide(organisms.get(j));
                }
            }
        }

        // remove dead organisms
        for (int i = organisms.size() - 1; i >= 0; i--)
        {
            if (!organisms.get(i).isAlive())
            {
                RemoveOrganism(organisms.get(i));
            }
        }


        AddLog("Turn " + ++turnNumber + " finished\n");
    }

    public ArrayList<String> getLogs()
    {
        return logs;
    }
    public void ClearLogs()
    {
        logs.clear();
    }
}

File: Project2/World.cpp
#include <conio.h>
#include "World.h"
#include <iostream>
#include <sstream>

World::World() : World::World(15, 15)
{
}
World::World(int height, int width)
{
	size = { height, width };
}
World::World(const World& orig)
{
	size = orig.size;
	turnNumber = orig.turnNumber;
	playerAlive = orig.playerAlive;
	organisms = std::vector<Organism*>(orig.organisms);
}
World::~World()
{
	for (size_t i = 0; i < organisms.size(); i++)
		delete organisms[i];
}

void World::AddOrganism(Organism* org)
{
	org->SetWorld(this);
	organisms.push_back(org);
}

void World::RemoveOrganism(Organism* org)
{
	for (size_t i = 0; i < organisms.size(); i++)
	{
		if (organisms[i] == org)
		{
			organisms.erase(organisms.begin() + i);
			delete org;
			break;
		}
	}
}

Organism* World::GetOrganism(const Vector2& pos) const
{
	for (size_t i = 0; i < organisms.size(); i++)
	{
		if (pos == organisms[i]->GetPosition())
			return organisms[i];
	}
	return nullptr;
}

std::vector<Organism*>& World::GetOrganisms()
{
	return organisms;
}

const Vector2& World::GetSize() const
{
	return size;
}

const size_t World::GetTurnNumber() const
{
	return turnNumber;
}

void World::SetTurnNumber(const size_t& value)
{
	turnNumber = value;
}

std::stringstream& World::Logs()
{
	return logs;
}

bool World::IsPlayerAlive() const
{
	return playerAlive;
}

void World::SetPlayerAlive(const bool& value)
{
	playerAlive = value;
}

bool World::ContainsPos(Vector2 pos) const
{
	return
		pos.x >= 0 && pos.x < size.x&&
		pos.y >= 0 && pos.y < size.y;
}

Vector2 World::GetRandomEmptyTile() const
{
	Vector2 pos = {};

	const int safetyLimit = 1000;
	for (int i = 0; i < safetyLimit; i++)
	{
		pos.x = rand() % size.x;
		pos.y = rand() % size.y;
		// if there is no organism on the chosen tile, return the position
		if (GetOrganism(pos) == nullptr)
			break;
	}

	return pos;
}

bool World::HasEmptyNeighbor(const Vector2 pos) const
{
	for (size_t i = 0; i < 4; i++)
	{
		Vector2 n = pos.GetNeighbor(i);
		if (ContainsPos(n) && GetOrganism(n) == nullptr)
			return true;
	}
	return false;
}

Vector2 World::GetEmptyNeighbor(const Vector2 pos) const
{
	std::vector<Vector2> targets;
	for (size_t i = 0; i < 4; i++)
	{
		Vector2 n = pos.GetNeighbor(i);
		if (ContainsPos(n) && GetOrganism(n) == nullptr)
			targets.push_back(n);
	}

	if (targets.size() > 0)
		return targets[rand() % targets.size()];
	else
		return { -1,-1 };
}

void World::ExecuteTurn()
{
	// order organisms by their priority
	struct ComparePriorities
	{
		bool operator() (const Organism* org1, const Organism* org2)
		{
			return (org1->GetPriority() > org2->GetPriority());
		}
	};
	std::stable_sort(organisms.begin(), organisms.end(), ComparePriorities());


	// execute actions
	size_t orgCount = organisms.size();
	for (size_t i = 0; i < orgCount; i++)
	{
		if (!organisms[i]->IsAlive())
			continue;

		//world->Logs() << i + 1 << ". ";
		organisms[i]->Action();
		
		// handle collisions
		for (size_t j = 0; j < organisms.size(); j++)
		{
			if (i != j && organisms[j]->IsAlive() && organisms[i]->GetPosition() == organisms[j]->GetPosition())
			{
				organisms[i]->Collide(organisms[j]);
			}
		}
	}

	// remove dead organisms
	for (int i = organisms.size() - 1; i >= 0; i--)
	{
		if (!organisms[i]->IsAlive())
		{
			RemoveOrganism(organisms[i]);
		}
	}


	logs << "Turn " << ++turnNumber << " finished\n";
}

void World::DrawWorld()
{
	// create a buffer for the drawing
	char** buffer = new char* [size.x];
	for (int x = 0; x < size.x; x++)
	{
		buffer[x] = new char[size.y + 1]();
		for (int y = 0; y < size.y; y++)
		{
			buffer[x][y] = ' ';
		}
	}

	// draw all organisms
	for (Organism* org : organisms)
	{
		org->Draw(buffer);
	}

	// draw the horizontal part of the frame
	_putch('#');
	for (int y = 0; y < size.y * 2 + 1; y++)
		_putch('-');
	_cputs("#\n");

	// draw the buffer on the console
	for (int x = 0; x < size.x; x++)
	{
		_putch('|');
		_putch(' ');
		for (int y = 0; y < size.y; y++)
		{
			_putch(buffer[x][y]);
			_putch(' ');
		}
		_cputs("|\n");
	}

	// draw the horizontal part of the frame
	_putch('#');
	for (int y = 0; y < size.y * 2 + 1; y++)
		_putch('-');
	_cputs("#\n");

	// deallocate the buffer
	for (int x = 0; x < size.x; x++)
	{
		delete[] buffer[x];
	}
	delete[] buffer;
}

void World::DrawLogs()
{
	std::cout << " --- Logs: ---\n\n" << logs.str();
}

void World::ClearLogs()
{
	logs.str("");
}

World& World::operator=(const World& other)
{
	World tmp(other);

	std::swap(size, tmp.size);
	std::swap(turnNumber, tmp.turnNumber);
	std::swap(playerAlive, tmp.playerAlive);
	std::swap(organisms, tmp.organisms);

	return *this;
}
